ql/src/java/org/apache/hadoop/hive/ql/ppd/OpProcFactory.java

```diff
@@ -417,9 +417,17 @@ public Object process(Node nd, Stack<Node> stack, NodeProcessorCtx procCtx,
         if (!ewi.isDeterministic()) {
           /* predicate is not deterministic */
           if (op.getChildren() != null && op.getChildren().size() == 1) {
+            ExprWalkerInfo prunedPreds = owi.getPrunedPreds((Operator<? extends OperatorDesc>) (op
+                .getChildren().get(0)));
+            //resolve of HIVE-23893
+            if (!(prunedPreds != null && prunedPreds.hasAnyCandidates())
```

Comment:
why the pruned predicates of the child could not be pushed down?

Reply:
in this condition we only process condition which contains both deterministic predicates and nondeterministic predicates, for other case, there is a fall through. you could notice that this condition is opposite to condition in createFilter.

Reply:
The fall through case contains the child's predicates which can be pushed down and they are the deterministic predicates. I think the goal is to push all the fittable predicates nearest to the top.

Reply:
There is a precondition that ewi.isDeterministic() is false, which means in this case there certainly contains nondeterministic predicates and may contains deterministic predicates. However,  we only concern there contains both nondeterministic predicates and  deterministic predicates.the fall through process case that child operator of current, not current.

Reply:
Like the case: select * from (select a.k as ak, b.* from a join b on a.k = b.k where rand(1) < 0.1 and a.hs = 11 and b.hs <= 10) x where ak = 'test';  the predicate ```ak = 'test'``` has been pushed to ``` rand(1) < 0.1 and a.hs = 11 and b.hs <= 10```,  the latter is non-deterministic, and ```!(prunedPreds != null && prunedPreds.hasAnyCandidates())``` is evaluated to false, so the fall though case is entered,  and the operation tree after this would be like ...-> filter(```rand(1) < 0.1 and a.hs = 11 and b.hs <= 10``` ) -> filter( ```ak = 'test'```) ->... , there may be seen no predicates pushed down in such cases.

Reply:
The perpose I add condition ``!(prunedPreds != null && prunedPreds.hasAnyCandidates())`` is that I do not want to destroy the original judgement, it's may could not do entirely satifactory, but it's sackless.The suggest you provide I will consider carefully, and I will think more situations.

Reply:
Tomorrow I will fix these bugs that found today, I will consider more test case tonight.

Reply:
I have remove redundant condition and take into account child preds which will be merge into new deterministic filter preds